lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1708,124 +1715,365 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
-      if (ty->is<MetatypeType>())
-        return false;
+      if (captureType->is<MetatypeType>()) {
+        requiresSelfQualification = false;
+      }
 
       // If self does not have reference semantics, it is very unlikely that
       // capturing it will create a reference cycle.
-      if (!ty->hasReferenceSemantics())
-        return false;
+      if (!captureType->hasReferenceSemantics()) {
+        requiresSelfQualification = false;
+      }
 
       if (auto closureExpr = dyn_cast<ClosureExpr>(inClosure)) {
-        if (auto selfDecl = closureExpr->getCapturedSelfDecl()) {
-          // If this capture is using the name `self` actually referring
-          // to some other variable (e.g. with `[self = "hello"]`)
-          // then implicit self is not allowed.
-          if (!selfDecl->isSelfParamCapture()) {
-            return true;
+        auto capturedSelfDecl = closureExpr->getCapturedSelfDecl();
+
+        // If this closure doesn't capture self explicitly, but this closure
+        // requires self qualification, then implicit self is disallowed.
+        if (!capturedSelfDecl && requiresSelfQualification) {
+          return false;
+        }
+
+        // If the closure has an explicit capture using the name `self` that
+        // actually refers to some other variable (e.g. `[self = "hello"]`)
+        // then implicit self is not allowed.
+        if (capturedSelfDecl && !isSimpleSelfCapture(capturedSelfDecl)) {
+          return false;
+        }
+      }
+
+      // If the self decl comes from a conditional statement, validate
+      // that it is an allowed `guard let self` or `if let self` condition.
+      //  - Even if this closure doesn't have a `weak self` capture, it could
+      //    be a closure nested in some parent closure with a `weak self`
+      //    capture, so we should always validate the conditional statement
+      //    that defines self if present.
+      if (validateSelfRebindings) {
+        if (auto conditionalStmt = parentConditionalStmt(selfDecl)) {
+          if (!hasValidSelfRebinding(conditionalStmt, inClosure)) {
+            return false;
           }
         }
       }
 
-      if (auto var = dyn_cast<VarDecl>(DRE->getDecl())) {
-        if (!isEnclosingSelfReference(var, inClosure)) {
+      // If this closure has a `weak self` capture, require that the
+      // closure unwraps self. If not, implicit self is not allowed
+      // in this closure or in any nested closure.
+      if (closureHasWeakSelfCapture(inClosure) &&
+          !hasValidSelfRebinding(parentConditionalStmt(selfDecl), inClosure)) {
+        return false;
+      }
+
+      if (auto autoclosure = dyn_cast<AutoClosureExpr>(inClosure)) {
+        // Implicit self is always allowed in autoclosure thunks generated
+        // during type checking. An example of this is when storing an instance
+        // method as a closure (e.g. `let closure = someInstanceMethodOnSelf`).
```

Comment:
I feel like this goes against the spirit of the proposal. Wouldn't you want to write `let closure = self.someInstanceMethod` for consistency and to make the self capture explicit? Or is this another historical quirk?

Reply:
Here's an example that fails to compile without this:

```swift
class Test {
  func foo() {
    let closure = someInstanceMethod
    closure()
  }
  
  func someInstanceMethod() { }
}
```

Allowing this to compile seems reasonable to me. It's also required for source compatibility.

However, if you instead access `someInstanceMethod` inside a closure, explicit self is required as you'd definitely expect:

```swift
class Test {
  func foo() {
    _ = {
      let closure = someInstanceMethod // error: error: unexpected error produced: implicit use of 'self' in closure
      closure()
    }
  }
  
  func someInstanceMethod() { }
}
```

I double-checked locally, the reason this works is that the implicit self decl is validated from both the context of the outer ClosureExpr and the AutoClosureExpr. This happens because the AST is `(ClosureExpr (DotSyntaxCallExpr (AutoClosureExpr (DeclRefExpr))))`, and both the DSCE and DRE are validated separately within the context of their containing closure using different rules.

Reply:
Yeah, doing the `self` capture analysis for bound, unapplied method references was discussed back in the SE-0269 days but never made its way into being an actual proposal. Not sure there'd be appetite for introducing _additional_ cases where implicit `self` would be disallowed at this point, though I'd probably count myself in favor :)